{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74835914",
   "metadata": {},
   "outputs": [],
   "source": [
    "#---------------------------------------Step 1.2---------------------------------------------\n",
    "import numpy as np\n",
    "from scipy.stats import poisson, norm\n",
    "\n",
    "genome_size = 1_000_000    # 1 Mbp\n",
    "coverage = 3      # 3× sequencing\n",
    "readlength = 100     # read length\n",
    "num_reads = (genome_size * coverage) // readlength\n",
    "\n",
    "genome_coverage = np.zeros(genome_size, dtype=int)\n",
    "\n",
    "rng = np.random.default_rng()\n",
    "for _ in range(num_reads):\n",
    "    startpos = rng.integers(0, genome_size - readlength + 1)\n",
    "    genome_coverage[startpos:startpos + readlength] += 1\n",
    "\n",
    "np.savetxt(\"coverage.txt\", genome_coverage)\n",
    "\n",
    "maxcoverage = int(genome_coverage.max())\n",
    "xs = np.arange(0, maxcoverage + 1)\n",
    "\n",
    "poisson_estimates = poisson.pmf(xs, mu=coverage)\n",
    "normal_estimates = norm.pdf(xs, loc=coverage, scale=np.sqrt(coverage))\n",
    "\n",
    "np.savetxt(\"coverage_values.txt\", xs)\n",
    "np.savetxt(\"poisson_estimates.txt\", poisson_estimates)\n",
    "np.savetxt(\"normal_estimates.txt\", normal_estimates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "192cb9c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#---------------------------------------Step 1.5---------------------------------------------\n",
    "import numpy as np\n",
    "from scipy.stats import poisson, norm\n",
    "\n",
    "genome_size = 1_000_000    # 1 Mbp\n",
    "coverage2 = 10      # 10× sequencing\n",
    "readlength2 = 100     # read length\n",
    "num_reads2 = (genome_size * coverage2) // readlength2\n",
    "\n",
    "genome_coverage2 = np.zeros(genome_size, dtype=int)\n",
    "\n",
    "rng2 = np.random.default_rng()\n",
    "for i in range(num_reads2):\n",
    "    startpos2 = rng2.integers(0, genome_size - readlength2 + 1)\n",
    "    genome_coverage2[startpos2:startpos2 + readlength2] += 1\n",
    "\n",
    "np.savetxt(\"coverage2.txt\", genome_coverage2)\n",
    "\n",
    "maxcoverage2 = int(genome_coverage2.max())\n",
    "xs2 = np.arange(0, maxcoverage2 + 1)\n",
    "\n",
    "poisson_estimates2 = poisson.pmf(xs2, mu=coverage2)\n",
    "normal_estimates2 = norm.pdf(xs2, loc=coverage2, scale=np.sqrt(coverage2))\n",
    "\n",
    "np.savetxt(\"coverage_values2.txt\", xs2)\n",
    "np.savetxt(\"poisson_estimates2.txt\", poisson_estimates2)\n",
    "np.savetxt(\"normal_estimates2.txt\", normal_estimates2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "70215a92",
   "metadata": {},
   "outputs": [],
   "source": [
    "#---------------------------------------Step 1.6---------------------------------------------\n",
    "import numpy as np\n",
    "from scipy.stats import poisson, norm\n",
    "\n",
    "genome_size = 1_000_000    # 1 Mbp\n",
    "coverage3 = 30      # 10× sequencing\n",
    "readlength3 = 100     # read length\n",
    "num_reads3 = (genome_size * coverage3) // readlength3\n",
    "\n",
    "genome_coverage3 = np.zeros(genome_size, dtype=int)\n",
    "\n",
    "rng3 = np.random.default_rng()\n",
    "for i in range(num_reads3):\n",
    "    startpos3 = rng3.integers(0, genome_size - readlength3 + 1)\n",
    "    genome_coverage3[startpos3:startpos3 + readlength3] += 1\n",
    "\n",
    "\n",
    "np.savetxt(\"coverage3.txt\", genome_coverage3)\n",
    "\n",
    "maxcoverage3 = int(genome_coverage3.max())\n",
    "xs3 = np.arange(0, maxcoverage3 + 1)\n",
    "\n",
    "poisson_estimates3 = poisson.pmf(xs3, mu=coverage3)\n",
    "normal_estimates3 = norm.pdf(xs3, loc=coverage3, scale=np.sqrt(coverage3))\n",
    "\n",
    "np.savetxt(\"coverage_values3.txt\", xs3)\n",
    "np.savetxt(\"poisson_estimates3.txt\", poisson_estimates3)\n",
    "np.savetxt(\"normal_estimates3.txt\", normal_estimates3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "568ee5aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "reads = ['ATTCA', 'ATTGA', 'CATTG', 'CTTAT', 'GATTG', 'TATTT', 'TCATT', 'TCTTA', 'TGATT', 'TTATT', 'TTCAT', 'TTCTT', 'TTGAT']\n",
    "k = 3\n",
    "\n",
    "graph = set()\n",
    "\n",
    "for read in reads:\n",
    "    for i in range(len(read) - k):\n",
    "        kmer1 = read[i:i+k]\n",
    "        kmer2 = read[i+1:i+1+k]\n",
    "        edge = f\"{kmer1} -> {kmer2}\"\n",
    "        graph.add(edge)\n",
    "\n",
    "with open(\"debruijn_graph.dot\", \"w\") as f:\n",
    "    f.write(\"digraph G {\\n\")\n",
    "    for edge in sorted(graph):\n",
    "        f.write(f\"    {edge};\\n\")\n",
    "    f.write(\"}\\n\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qb25",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
